File: frontend/src/utils/budgetCalculator.ts
const FEDERAL_BRACKETS = [
  { rate: 0.10, min: 0, max: 11925 },
  { rate: 0.12, min: 11925, max: 48475 },
  { rate: 0.22, min: 48475, max: 103350 },
  { rate: 0.24, min: 103350, max: 197300 },
  { rate: 0.32, min: 197300, max: 250525 },
  { rate: 0.35, min: 250525, max: 626350 },
  { rate: 0.37, min: 626350, max: Infinity },
]

const STANDARD_DEDUCTION = 14600
const SS_WAGE_BASE = 176100
const SS_RATE = 0.062
const MEDICARE_RATE = 0.0145

export const STATE_TAX_RATES: Record<string, number> = {
  AL: 0.05, AK: 0, AZ: 0.025, AR: 0.044, CA: 0.093,
  CO: 0.044, CT: 0.065, DE: 0.066, FL: 0, GA: 0.055,
  HI: 0.079, ID: 0.058, IL: 0.0495, IN: 0.03, IA: 0.06,
  KS: 0.057, KY: 0.04, LA: 0.042, ME: 0.075, MD: 0.0575,
  MA: 0.05, MI: 0.0425, MN: 0.0785, MS: 0.047, MO: 0.054,
  MT: 0.059, NE: 0.0664, NV: 0, NH: 0, NJ: 0.0637,
  NM: 0.059, NY: 0.0685, NC: 0.0475, ND: 0.029, OH: 0.04,
  OK: 0.0475, OR: 0.099, PA: 0.0307, RI: 0.0599, SC: 0.064,
  SD: 0, TN: 0, TX: 0, UT: 0.0465, VT: 0.0666,
  VA: 0.0575, WA: 0, WV: 0.065, WI: 0.0765, WY: 0, DC: 0.085,
}

export const US_STATES = [
  { code: 'AL', name: 'Alabama' }, { code: 'AK', name: 'Alaska' },
  { code: 'AZ', name: 'Arizona' }, { code: 'AR', name: 'Arkansas' },
  { code: 'CA', name: 'California' }, { code: 'CO', name: 'Colorado' },
  { code: 'CT', name: 'Connecticut' }, { code: 'DE', name: 'Delaware' },
  { code: 'FL', name: 'Florida' }, { code: 'GA', name: 'Georgia' },
  { code: 'HI', name: 'Hawaii' }, { code: 'ID', name: 'Idaho' },
  { code: 'IL', name: 'Illinois' }, { code: 'IN', name: 'Indiana' },
  { code: 'IA', name: 'Iowa' }, { code: 'KS', name: 'Kansas' },
  { code: 'KY', name: 'Kentucky' }, { code: 'LA', name: 'Louisiana' },
  { code: 'ME', name: 'Maine' }, { code: 'MD', name: 'Maryland' },
  { code: 'MA', name: 'Massachusetts' }, { code: 'MI', name: 'Michigan' },
  { code: 'MN', name: 'Minnesota' }, { code: 'MS', name: 'Mississippi' },
  { code: 'MO', name: 'Missouri' }, { code: 'MT', name: 'Montana' },
  { code: 'NE', name: 'Nebraska' }, { code: 'NV', name: 'Nevada' },
  { code: 'NH', name: 'New Hampshire' }, { code: 'NJ', name: 'New Jersey' },
  { code: 'NM', name: 'New Mexico' }, { code: 'NY', name: 'New York' },
  { code: 'NC', name: 'North Carolina' }, { code: 'ND', name: 'North Dakota' },
  { code: 'OH', name: 'Ohio' }, { code: 'OK', name: 'Oklahoma' },
  { code: 'OR', name: 'Oregon' }, { code: 'PA', name: 'Pennsylvania' },
  { code: 'RI', name: 'Rhode Island' }, { code: 'SC', name: 'South Carolina' },
  { code: 'SD', name: 'South Dakota' }, { code: 'TN', name: 'Tennessee' },
  { code: 'TX', name: 'Texas' }, { code: 'UT', name: 'Utah' },
  { code: 'VT', name: 'Vermont' }, { code: 'VA', name: 'Virginia' },
  { code: 'WA', name: 'Washington' }, { code: 'WV', name: 'West Virginia' },
  { code: 'WI', name: 'Wisconsin' }, { code: 'WY', name: 'Wyoming' },
  { code: 'DC', name: 'Washington, D.C.' },
]

function calcFederalTax(taxableIncome: number): number {
  if (taxableIncome <= 0) return 0
  let tax = 0
  for (const bracket of FEDERAL_BRACKETS) {
    if (taxableIncome <= bracket.min) break
    const taxableInBracket = Math.min(taxableIncome, bracket.max) - bracket.min
    tax += taxableInBracket * bracket.rate
  }
  return tax
}

export type PayFrequency = 'weekly' | 'biweekly' | 'semimonthly' | 'monthly'

export interface DirectDepositInput {
  mode: 'direct'
  takeHome: number
  frequency: PayFrequency
  savingsRate: number
}

export interface SalaryInput {
  mode: 'salary'
  grossAnnual: number
  state: string
  retirement401kPct: number
  monthlyHealthcare: number
  savingsRate: number
}

export interface BudgetCalculationResult {
  recommendedMonthlyBudget: number
  breakdown?: {
    grossAnnual: number
    retirement401k: number
    healthcare: number
    federalTax: number
    stateTax: number
    fica: number
    netAnnual: number
    savingsAmount: number
  }
}

const FREQUENCY_MULTIPLIERS: Record<PayFrequency, number> = {
  weekly: 52,
  biweekly: 26,
  semimonthly: 24,
  monthly: 12,
}

export function calculateBudget(input: DirectDepositInput | SalaryInput): BudgetCalculationResult {
  if (input.mode === 'direct') {
    const annualTakeHome = input.takeHome * FREQUENCY_MULTIPLIERS[input.frequency]
    const monthlyTakeHome = annualTakeHome / 12
    const recommendedMonthlyBudget = monthlyTakeHome * (1 - input.savingsRate / 100)
    return { recommendedMonthlyBudget }
  }

  const { grossAnnual, state, retirement401kPct, monthlyHealthcare, savingsRate } = input

  const retirement401k = grossAnnual * (retirement401kPct / 100)
  const annualHealthcare = monthlyHealthcare * 12

  // FICA on gross (SS capped at wage base, Medicare uncapped)
  const ssWages = Math.min(grossAnnual, SS_WAGE_BASE)
  const ssTax = ssWages * SS_RATE
  const medicareTax = grossAnnual * MEDICARE_RATE
  const fica = ssTax + medicareTax

  // Federal taxable income: gross - 401k - standard deduction
  const federalTaxable = Math.max(0, grossAnnual - retirement401k - STANDARD_DEDUCTION)
  const federalTax = calcFederalTax(federalTaxable)

  // State tax on gross (approximate — most states don't allow 401k deduction in same way)
  const stateRate = STATE_TAX_RATES[state] ?? 0
  const stateTax = grossAnnual * stateRate

  const netAnnual = grossAnnual - retirement401k - annualHealthcare - federalTax - stateTax - fica
  const monthlyNet = netAnnual / 12
  const savingsAmount = monthlyNet * (savingsRate / 100)
  const recommendedMonthlyBudget = monthlyNet * (1 - savingsRate / 100)

  return {
    recommendedMonthlyBudget: Math.max(0, recommendedMonthlyBudget),
    breakdown: {
      grossAnnual,
      retirement401k,
      healthcare: annualHealthcare,
      federalTax,
      stateTax,
      fica,
      netAnnual: Math.max(0, netAnnual),
      savingsAmount: Math.max(0, savingsAmount),
    },
  }
}

File: ios/BudgetMasterApp/Budget Chat/Utilities/BudgetCalculator.swift
import Foundation

// MARK: - PayFrequency

public enum PayFrequency: String, CaseIterable, Identifiable {
    case weekly = "Weekly"
    case biweekly = "Biweekly"
    case semimonthly = "Semi-monthly"
    case monthly = "Monthly"

    public var id: String { rawValue }

    public var periodsPerYear: Double {
        switch self {
        case .weekly:      return 52
        case .biweekly:    return 26
        case .semimonthly: return 24
        case .monthly:     return 12
        }
    }
}

// MARK: - IncomeMode

public enum IncomeMode: String, CaseIterable {
    case directDeposit = "I know my take-home"
    case salary = "Calculate from salary"
}

// MARK: - BudgetCalcInput

public struct BudgetCalcInput {
    public var mode: IncomeMode = .directDeposit
    // Direct deposit fields
    public var takeHome: Double = 0
    public var frequency: PayFrequency = .biweekly
    // Salary fields
    public var grossAnnual: Double = 0
    public var stateCode: String = "CA"
    public var retirement401kPct: Double = 0    // 0–100
    public var monthlyHealthcare: Double = 0
    // Shared
    public var savingsRate: Double = 20         // 0–100

    public init() {}
}

// MARK: - BudgetCalcResult

public struct BudgetCalcResult {
    public let recommendedMonthlyBudget: Double
    public let breakdown: BudgetBreakdown?
}

// MARK: - BudgetBreakdown

public struct BudgetBreakdown {
    public let grossAnnual: Double
    public let retirement401k: Double
    public let healthcare: Double
    public let federalTax: Double
    public let stateTax: Double
    public let fica: Double
    public let netAnnual: Double
    public let savingsAmount: Double
}

// MARK: - BudgetCalculator

public struct BudgetCalculator {

    // MARK: State Tax Rates (2025 Tax Foundation data, effective rate estimates)

    /// Flat or estimated effective marginal rates for all 50 states + DC.
    /// No-income-tax states are 0.0.
    private static let stateTaxRates: [String: Double] = [
        // No-income-tax states
        "AK": 0.000,
        "FL": 0.000,
        "NV": 0.000,
        "NH": 0.000,
        "SD": 0.000,
        "TN": 0.000,
        "TX": 0.000,
        "WA": 0.000,
        "WY": 0.000,
        // Flat-tax states
        "IL": 0.0495,
        "PA": 0.0307,
        "IN": 0.0300,
        "KY": 0.0400,
        "CO": 0.0440,
        "MI": 0.0425,
        "UT": 0.0465,
        "NC": 0.0475,
        // Progressive states — estimated effective rates at moderate income
        "CA": 0.0930,
        "NY": 0.0685,
        "OR": 0.0990,
        "MN": 0.0785,
        "HI": 0.0820,
        "NJ": 0.0637,
        "VT": 0.0660,
        "DC": 0.0850,
        "ME": 0.0715,
        "SC": 0.0700,
        "CT": 0.0699,
        "WI": 0.0765,
        "ID": 0.0580,
        "MT": 0.0590,
        "NE": 0.0664,
        "ND": 0.0290,
        "RI": 0.0599,
        "MA": 0.0500,
        "OH": 0.0399,
        "GA": 0.0550,
        "VA": 0.0575,
        "MD": 0.0575,
        "AL": 0.0500,
        "MS": 0.0500,
        "LA": 0.0425,
        "AR": 0.0490,
        "MO": 0.0495,
        "KS": 0.0570,
        "OK": 0.0475,
        "AZ": 0.0250,
        "NM": 0.0490,
        "WV": 0.0650,
        "DE": 0.0660,
        "IA": 0.0600,
    ]

    // MARK: Federal Tax Brackets (2025, single filer, marginal)

    // Brackets are (upper bound, rate). The last bracket has no upper bound
    // (represented by .infinity).
    private static let federalBrackets: [(Double, Double)] = [
        (11600,    0.10),
        (47150,    0.12),
        (100525,   0.22),
        (191950,   0.24),
        (243725,   0.32),
        (609350,   0.35),
        (.infinity, 0.37),
    ]

    // MARK: Constants

    /// 2025 Social Security wage base
    private static let ssWageBase: Double = 176_100
    /// 2025 standard deduction for single filers
    private static let standardDeduction: Double = 14_600

    // MARK: Public API

    public static func calculateBudget(input: BudgetCalcInput) -> BudgetCalcResult {
        switch input.mode {
        case .directDeposit:
            return calculateDirectDeposit(input: input)
        case .salary:
            return calculateFromSalary(input: input)
        }
    }

    // MARK: - Direct Deposit Path

    private static func calculateDirectDeposit(input: BudgetCalcInput) -> BudgetCalcResult {
        let annualTakeHome = input.takeHome * input.frequency.periodsPerYear
        let monthlyNet = annualTakeHome / 12.0
        let monthlyBudget = monthlyNet * (1.0 - input.savingsRate / 100.0)
        return BudgetCalcResult(recommendedMonthlyBudget: monthlyBudget, breakdown: nil)
    }

    // MARK: - Salary Path

    private static func calculateFromSalary(input: BudgetCalcInput) -> BudgetCalcResult {
        let gross = input.grossAnnual

        // Pre-tax deductions
        let retirement401k = gross * (input.retirement401kPct / 100.0)
        let healthcareAnnual = input.monthlyHealthcare * 12.0

        // Taxable income for federal purposes
        let taxableIncome = max(0, gross - retirement401k - healthcareAnnual - standardDeduction)

        // Federal tax via marginal brackets
        let federalTax = marginalTax(on: taxableIncome, brackets: federalBrackets)

        // State tax (flat effective rate on gross income, approximation)
        let stateRate = stateTaxRates[input.stateCode.uppercased()] ?? 0.05
        let stateTax = gross * stateRate

        // FICA: Social Security (6.2% up to wage base) + Medicare (1.45% uncapped)
        let fica = min(gross, ssWageBase) * 0.062 + gross * 0.0145

        // Net annual take-home
        let netAnnual = gross - retirement401k - healthcareAnnual - federalTax - stateTax - fica

        // Monthly spendable after savings
        let monthlyNet = netAnnual / 12.0
        let savingsAmount = monthlyNet * (input.savingsRate / 100.0)
        let monthlyBudget = monthlyNet * (1.0 - input.savingsRate / 100.0)

        let breakdown = BudgetBreakdown(
            grossAnnual: gross,
            retirement401k: retirement401k,
            healthcare: healthcareAnnual,
            federalTax: federalTax,
            stateTax: stateTax,
            fica: fica,
            netAnnual: max(0, netAnnual),
            savingsAmount: savingsAmount
        )

        return BudgetCalcResult(
            recommendedMonthlyBudget: max(0, monthlyBudget),
            breakdown: breakdown
        )
    }

    // MARK: - Tax Helpers

    private static func marginalTax(on income: Double, brackets: [(Double, Double)]) -> Double {
        var tax: Double = 0
        var previousBound: Double = 0

        for (upperBound, rate) in brackets {
            if income <= previousBound { break }
            let taxableInBracket = min(income, upperBound) - previousBound
            tax += taxableInBracket * rate
            previousBound = upperBound
        }

        return tax
    }
}

// MARK: - US States List (for picker)

public struct USState: Identifiable {
    public let code: String
    public let name: String
    public var id: String { code }
}

public let allUSStates: [USState] = [
    USState(code: "AL", name: "Alabama"),
    USState(code: "AK", name: "Alaska"),
    USState(code: "AZ", name: "Arizona"),
    USState(code: "AR", name: "Arkansas"),
    USState(code: "CA", name: "California"),
    USState(code: "CO", name: "Colorado"),
    USState(code: "CT", name: "Connecticut"),
    USState(code: "DC", name: "Washington D.C."),
    USState(code: "DE", name: "Delaware"),
    USState(code: "FL", name: "Florida"),
    USState(code: "GA", name: "Georgia"),
    USState(code: "HI", name: "Hawaii"),
    USState(code: "ID", name: "Idaho"),
    USState(code: "IL", name: "Illinois"),
    USState(code: "IN", name: "Indiana"),
    USState(code: "IA", name: "Iowa"),
    USState(code: "KS", name: "Kansas"),
    USState(code: "KY", name: "Kentucky"),
    USState(code: "LA", name: "Louisiana"),
    USState(code: "ME", name: "Maine"),
    USState(code: "MD", name: "Maryland"),
    USState(code: "MA", name: "Massachusetts"),
    USState(code: "MI", name: "Michigan"),
    USState(code: "MN", name: "Minnesota"),
    USState(code: "MS", name: "Mississippi"),
    USState(code: "MO", name: "Missouri"),
    USState(code: "MT", name: "Montana"),
    USState(code: "NE", name: "Nebraska"),
    USState(code: "NV", name: "Nevada"),
    USState(code: "NH", name: "New Hampshire"),
    USState(code: "NJ", name: "New Jersey"),
    USState(code: "NM", name: "New Mexico"),
    USState(code: "NY", name: "New York"),
    USState(code: "NC", name: "North Carolina"),
    USState(code: "ND", name: "North Dakota"),
    USState(code: "OH", name: "Ohio"),
    USState(code: "OK", name: "Oklahoma"),
    USState(code: "OR", name: "Oregon"),
    USState(code: "PA", name: "Pennsylvania"),
    USState(code: "RI", name: "Rhode Island"),
    USState(code: "SC", name: "South Carolina"),
    USState(code: "SD", name: "South Dakota"),
    USState(code: "TN", name: "Tennessee"),
    USState(code: "TX", name: "Texas"),
    USState(code: "UT", name: "Utah"),
    USState(code: "VT", name: "Vermont"),
    USState(code: "VA", name: "Virginia"),
    USState(code: "WA", name: "Washington"),
    USState(code: "WV", name: "West Virginia"),
    USState(code: "WI", name: "Wisconsin"),
    USState(code: "WY", name: "Wyoming"),
]
